tron/core/recovery.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import logging
+
+from tron.actioncommand import NoActionRunnerFactory
+from tron.core.actionrun import ActionRun
+from tron.core.actionrun import SSHActionRun
+
+log = logging.getLogger(__name__)
+
+
+def filter_action_runs_needing_recovery(action_runs):
+    return [
+        action_run for action_run in action_runs
+        if action_run.state == ActionRun.STATE_UNKNOWN
+    ]
+
+
+def filter_recoverable_action_runs(action_runs):
+    """
+    Given a list of action_runs, create a filtered list that only includes those that can be recovered
+    For now, the only test is whether the run is an SSHActionRun
+    """
+    return [
+        action_run for action_run in action_runs
+        if isinstance(action_run, SSHActionRun)
+    ]
+
+
+def filter_recovery_candidates(runs):
+    return filter_recoverable_action_runs(
+        action_runs=filter_action_runs_needing_recovery(action_runs=runs),
+    )
+
+
+def build_recovery_command(recovery_binary, path):
+    return "%s %s" % (recovery_binary, path)
+
+
+def recover_action_run(action_run, action_runner):
+    log.info("creating recovery run for actionrun %s" % action_run.id)
+    if type(action_runner) == NoActionRunnerFactory:
+        log.info(
+            "unable to recover action_run %s: action_run has no action_runner"
+            % action_run.id,
+        )
+        return None
+
+    recovery_run = SSHActionRun(
+        job_run_id=action_run.job_run_id,
+        name="recovery-%s" % action_run.id,
```

Comment:
if this object is supposed to be a drop-in replacement for original action run object, are we sure nothing depends on `name` attribute being the original action name?

Reply:
afaik the name field is just a vanity field, but I'll double check.